### ISSUE TITLE
Feature Request: add support for innerloop Time-Travel Debugging

### ISSUE THREAD

**Issue by mrkmarron** [Microsoft/vscode#42383]
This request builds on the following work that was done earlier to support TTD for Node. 

We have implemented preliminary support for innerloop time-travel debugging using the multi-target debugging functionality in VSCode [demo gif](https://1drv.ms/i/s!AgleYbnkXJEogfxJoMbb7NQs_vzfVg). 
The design we went with emphasizes a general workflow that can support a range of time-travel debuggers for languages other that Node.

Most of the coding has been completed and PR's are open:
- [chrome-debug-core](https://github.com/Microsoft/vscode-chrome-debug-core/pull/275)
- [node-debug-2](https://github.com/Microsoft/vscode-node-debug2/pull/164)
- [vscode](https://github.com/Microsoft/vscode/pull/42384)

Modulo feedback on PR's we have the following open issues:
- [ ] Add loading animation/status/info and cancel option on launch time-travel. This may take a few seconds to complete so we want to provide some feedback.
- [ ] Figure out what should happen if a TT session is in progress and a second one is requested.
- [ ] Workout possible temp directory approach in debugger extension.
- [ ] Needs code to clean/create location for storing a trace.
- [ ] Start TTD icon needs a better version.

**Comment by weinand** [comment]
@mrkmarron thanks for this extensive contribution. In general we prefer to receive a feature request first and then start a discussion before we might consider accepting external contributions.

So, before we can start reviewing PRs, we'd like to understand the scope of the feature and the architectural implications on both VS Code and the Debug Adapter Protocol.
 
I suggest that you first explain (as bullet-items) what is needed for "innerloop Time-Travel Debugging" and what is missing from VS Code and the Debug Adapter Protocol.

**Comment by mrkmarron** [comment]
Hi @weinand, I had a brief discussion with Rob (@roblourens) last week before I started coding this but happy to expand on that now that there is more concrete info. The scope of the feature is to enable debugger support for innerloop time-travel debugging for (Node via the builtin node2 extension in this commit) but also debugger extension providers for other languages going forward. 

The proposed approach to implementing this feature leverages the existing multi-target debugging to let VSCode launch a "replay debug session" from a "live ttd session" via a new UI element in the debug bar (logic for this UI element) and new methods in the IDebugService interface (although we may want to split this out into an ITTDebugService interface). The changes to the node2 and chrome-debug implement the needed functionality for Node when a TTD enabled version is detected (Node-ChakraCore). All of these changes sit behind the TTD enabled checks so only change functionality in time-travel debugging flows and have minimal impacts on any other execution.

Overall this changes is structured to minimize impact on the current debugger flow while:
1. Immediately enabling us to provide innerloop TTD for Node-ChakraCore which is an item that @bterlson and I have been discussing and @aruneshchandra and I frequently hear as a question.
2. Providing a simple structure that other TTD systems can easily use to provide innerloop TTD for their target language. 

# VSCode:
1. Add StartTTDAction UI element and action (debugAction.ts, widget, and CSS files):
    - Checks if the current debugger supports TTD and allows for innerloop debugging (newly added isTTDLiveMode interface method).
    - We need to get a location to write out the time-travel trace (newly added getTTDTraceWriteURI interface method).
    - Then ask the debugger to write this trace out to the new URI (newly added writeTTDLog method) after we have ensured it exists/is empty (ensureDirectoryAndEmpty not implemented currently).
    - Finally we ask the debugger to generate the appropriate launch config for replay-debugging of this trace (newly added getTTDReplayConfiguration interface method) and then launch it.
  - As noted we probably want some nicer waiting/activity notification during this process.
2. Add interface methods that the debug adapters need to implement to support this, provide signatures, and plumb things through (debug.ts, debugModel.ts, debugProtocol.d.ts, rawDebugSession.ts)
    - The current implementation places them in IDebugService but there may be other alternatives

# node-debug2:
1. We add implementations of the new interface methods getTTDTraceWriteURI, getTTDReplayConfiguration, isTTDLiveMode, and isTTDReplayMode
    - getTTDReplayConfiguration, isTTDLiveMode, and isTTDReplayMode are all simple processing of known Node configuration arguments
    - getTTDTraceWriteURI currently uses a quick-dirty directory construction so I would like to review options for this (in particular using temp location)
 
# chrome-debug:
1. The required change is plumbing through the call to writeTTDLog and providing defaults for the other interface methods. 
    - We also have a helper for cleaning up the trace write location and do that here as a temporary? workaround. I believe this code should be in the VSCode component but would like to feedback on this.

**Comment by weinand** [comment]
@mrkmarron thanks, this information is a start point. 

After we have finished the current January endgame I will have time to digest the information and will come back to you with questions...

Here are already two questions:
- VS Code communicates with debug adapters via the Debug Adapter Protocol (DAP). Before we implement anything for the VS Code debugger UI or Debug Adapters (DA) we need a detailed specification of all required DAP additions. Please note that additions to the DAP do not happen lightly. There needs to be agreement that DAP additions are useful and can be implemented by clients (e.g. VS, VS Code, VS for Mac) as well as debug extensions. Typically DAP discussions take time (~ 2-3 milestones). Here is a good example what we expect and how the resulting discussion looks like: https://github.com/Microsoft/vscode-debugadapter-node/issues/108.
- did you try to implement your feature as an extension? Why not? What are the limitations of the VS Code debugger API? A good approach for adding new debugger features to VS Code is by starting with an extension first. Initially the extension can use private DAP requests to talk to the DA. If the extension works and is successful, the private DAP can be converted into official DAP. This makes it possible that DAs can already start to implement it. In a second step the feature's UI can be moved from an extension to VS Code's core. The "Loaded Scripts" feature from https://github.com/Microsoft/vscode-debugadapter-node/issues/108 is currently in the middle of this process: the private protocol has been made official in the DAP and DAs are implementing it. But the VS Code UI implementation is still waiting for being migrated into VS Code.  


The typical way to extend VS Code is by writing an extension. Have you tried this?

**Comment by weinand** [comment]
@mrkmarron Since we just finished the February release, I had a chance to do more thorough review of this feature request and the related PRs.

Here is my feedback:

We always try to keep debugger specific code outside of VS Code because this code becomes easily broken by an evolving debugger and we cannot maintain this code easily if we are not owning the corresponding debugger ourselves.

For that reason I do not like to have the `StartTTDAction` and the code that deals with "TTDLogs" in the VS Code Core. This code is just too specific (and it's not even complete yet).
I suggest that you move this functionality into a "TTD" extension. We can decide later whether to merge this into the node-debug or chrome-debug extensions so that a user does not have to install an additional extension explicitely.
But for now an extension is the best way to decouple things and make progress.

A "TTD" extension can easily contribute a `StartTTDAction` that a user can find via F1 or trigger with a keyboard shortcut.

An extension cannot contribute actions to the debug action bar, so you won't be able to show the "StartTTDAction" there. But I didn't like the TTD action anyway because the debug action bar is not a general toolbar. Its purpose is "stepping", and "starting a TTD session" does not belong there.

The implementation of the "StartTTDAction" action can communicate via a `DebugSession.customRequest` and `debug.onDidReceiveDebugSessionCustomEvent` with the underlying debug session, retrieve a TTDLog, create a new launch config in memory and launch the TTD debug session via `vscode.debug.StartDebugging(...)`. In addition extension API exists for showing progress.

Here is just a (naive) sketch for the `StartTTDAction`:
```ts
context.subscriptions.push(vscode.commands.registerCommand('extension.startTTD', async () => {

const session = vscode.debug.activeDebugSession;
if (session && session.type === 'node2') {

    // Replace 'evaluate' with DAP requests that retrieve TTDLog etc.
    const reply = await session.customRequest('evaluate', { expression: '"uri"' });
    if (reply.result) {

    // launch TTD session
    vscode.debug.startDebugging(undefined, {
        type: "node",
        request: "launch",
        name: "TTDReplay",
        protocol: "inspector",
        runtimeArgs: [`--replay-debug=${reply.result}`],
        console: "internalConsole"
    });

    } else {
    // can't retrieve TDDLogs
    }
} else {
    // no active node2 session
}
}));
```

What is missing from your PRs:
- multi session support: a user can have multiple debug sessions that support TTD. You need to manage a robust association between the node/chrome session and the corresponding TTD session and its TTDLogs.
- cleaning up TTDLogs.
- supporting multi-root workspaces

The missing things are another argument for implementing TTD as an extension: you can fill the gaps in a piecemeal way and don't have to wait until we have reviewed and merged any additional PRs.

What do you think?

**Comment by mrkmarron** [comment]
Based on this discussion I have heavily refactored the code to work almost entirely as an extension. However, a small change (PR #46441)[https://github.com/Microsoft/vscode/pull/46441] is needed in the TTD stepping specific code to support this implementation.

**Comment by weinand** [comment]
@mrkmarron since we have merged [PR #46441](https://github.com/Microsoft/vscode/pull/46441) is it OK if we close this issue?
